### DISCUSSION TITLE
Hybrid X509 Certificates - Clarification on the produced content

### DISCUSSION THREAD

**Discussion by dimhatzi** [open-quantum-safe/openssl#419] [General]
Dear all,
I would kindly requesting your valuable feedback related to the creation of hybrid certificates for digital signatures.
More specifically, by reading the information related to the Section of the Supported Algorithms for Authentication, I read the following:

"if SIG claims NIST L1 or L2 security, then the fork provides the methods rsa3072_SIG and p256_SIG, **which combine** SIG with RSA3072 and with ECDSA using NIST's P256 curve respectively."

Is there any way to seek more information how this actually works and what the word "**_combine_**" means? Is it about a serial concatenation of the public keys and the signature values, is it a kind of encapsulation or anything else?

I really appreciate your time and valuable feedback.

Kind regards

**Comment by baentsch** [comment]
> Is there any way to seek more information how this actually works and what the word "combine" means?

It's concatenation. Depending on your preferences, you may want to read either a paper: https://s3.amazonaws.com/files.douglas.stebila.ca/files/research/papers/NISTPQC-CroPaqSte19.pdf or the code, e.g. https://github.com/open-quantum-safe/openssl/blob/38e12f18693b6145df7c955bb70c77d2a6e0f164/crypto/ec/oqs_meth.c#L990-L1049 showing print-out of hybrid keys (in general, check the code for the `is_hybrid` flag).

**Reply by dimhatzi** [reply]
Thank you for the quick reply.
And how is it possible to differentiate the concatenated values, since I expect that there will not be only the pure key data and signature values data concatenated.
For example, I have produced a trust chain (RootCA and IntermediateCA) for the creation of a hybrid user certificate using rsa3072_dilithium2_aes.

I have the below findings and I would appreciate your advice:
1. Public Keys Section
When I read the x509 using openSSL command, the depiction of the public key values (for both RSA and Dilithium) are fine. But when serialised, indeed they are concatenated, but at the start and in-between the concatenated values, there are some values, which I suppose have to do with some additional information about the public keys. More specifically, the following bytes where created:
**00 00 01 8E 30 82 01 8A 02 82 01 81** {+RSA Public Key} **02 03 01 00 01** {+dilithium2_aes Public Key}
The first octet string is unknown to me. The second one (after a HEX to ASN.1convertion) gives the value of _U.P.INTEGER 0x010001 (65537 decimal)_, which I suppose it corresponds to the RSA "Exponent: 65537 (0x10001)" value of the produced certificate.
I also want to note that the public key length of the RSA is 385 and not the expected 384.

2. Signature Value Section
The total length of the produced concatenated value is 2808 bytes.
If I am not mistaken, the signature value length of the rsa3072 is 384 bytes and the length of the signature value of dilithium2_aes is 2420. The sum makes 2804 bytes. I am missing 4 bytes and I do not know how to locate them in order to know the exact signature values for each algorithm.

Your advice and help is much appreciated :)

Kind regards

**Reply by baentsch** [reply]
The above seems to point to some encoding (padding?) issues, so my best suggestion is to read through the X.509 specification to get an explanation as the logic generating these structures is unmodified OpenSSL (X.509) code.

**Reply by dimhatzi** [reply]
Thank you. Indeed this is the case.

**Comment by pfuentes69** [comment]
Hello,
is there any way to generate hybrid certificates with nested signatures?
I mean using the "Alt Signature" extensions.
Txs!

**Reply by baentsch** [reply]
> Hello, is there any way to generate hybrid certificates with nested signatures? 

With near-certainty the answer is "No", but to be sure:

> I mean using the "Alt Signature" extensions. Txs!

Could you please point to the spec/RFC that you mean?

**Reply by pfuentes69** [reply]
Hello,
You can check section 9.8 (Page 58) of  ITU-T X.509 | ISO/IEC 9594-8 (https://handle.itu.int/11.1002/1000/14033-en?locatt=format:pdf&auth)

You can see an example in the attached cert.
Txs

[rsa_hss_root_private.cert.pem.zip](https://github.com/open-quantum-safe/openssl/files/10686501/rsa_hss_root_private.cert.pem.zip)

**Reply by baentsch** [reply]
Any input as to how [this RFC](https://datatracker.ietf.org/doc/html/draft-truskovsky-lamps-pq-hybrid-x509-00) relates to your question?

**Reply by pfuentes69** [reply]
Well, that RFC defines the same alt signature extensions, so my question keeps being if OQS supports this, as it seems the path the industry is following and what other quantum safe implementations support already

**Reply by baentsch** [reply]
The answer is "No" for oqs-openssl111 and "not yet" for oqs-provider (OpenSSL3): A person indicated interest to contribute such code to that OQS software but without promising a date.
